### PR TITLE
Deserialize JSON using KotlinX Serialization instead of Gson

### LINKED ISSUE #224: Migrate from gson to KotlinX serialization
According to the link below, gson is in maintenance mode. There will be bug fixes but no new feature development. Therefore, Gson should be replaced with something actively maintained. KotlinX serialization was proposed by @PattaFeuFeu , which is a nice fit for this Kotlin library.

https://github.com/google/gson

### PR DESCRIPTION
**Pull request by PattaFeuFeu**
# Description

In this PR, I replace GSON as the JSON (de-)serialiser of choice with [KotlinX’ Serialization](https://github.com/Kotlin/kotlinx.serialization) library.

#224 mentioned Jackson as one of the options but KotlinX Serialization seems like a better fit to me with it being a KotlinX library and thus hopefully even better support and more “kotlin-y” APIs.
The Java usage sometimes gets a bit clunky but that shouldn’t be an issue as we don’t expose the serializer. **I have made it an internal dependency**, as opposed to GSON which was an `api` instead of `implementation` gradle dependency.

The usage of GSON in this project was pretty basic but if we need more advanced customer deserializers, this [Medium article about this exact migration](https://medium.com/livefront/migration-gson-to-kotlinx-serialization-ff43cd03631a) could come in handy.

# Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] Dependency replacement

# How Has This Been Tested?

I ran `gradle check` and tested the Kotlin and Java sample app with a test instance

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation if necessary
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

---

**History of this description as some comments already referenced the old state:**

This PR should in the end close #224.

## State of this PR

This PR is not yet entirely ready for review. It’s just a first work in progress that I’d like your feedback on. @andregasser 

1. Is this a good way forward?
2. Is my change in the test JSON data from Int to String for id-related properties okay? (see e.g. [Status API doc](https://docs.joinmastodon.org/entities/Status/) which mentions “Type: String (cast from an integer but not guaranteed to be a number)”
3. Are the code-style related changes okay?

### Code style

When I applied code style settings from this project in IntelliJ IDEA, auto-formatting changed loads of lines so I assume there’s no style added to the project? I went for the Kotlin defaults then which seemed to mostly not change too much. I did not merge auto-formatting changes unrelated to this PR’s main changes though.

Usually, I use argument names heavily but this project doesn’t seem to do that. Was that a conscious choice? If so, I’m glad to remove the few additions again. If not, I’ll keep them. 😊 

Same for `var`/`val` types and `fun` return types. I usually explicitly add them as well but also couldn’t see that everywhere here.

## Still open

~~I needed to litter the JSON config in a few different files and for finishing up I’d like to find a better approach here that would still ensure that we’re using the same config everywhere without adding too much coupling. I already don’t like the JSON serializer being part of `MastododonClient` but I wanted to keep this PR as close to the original GSON usage as possible _for now_.~~

### CODE REVIEW

**Comment by andregasser** [comment]
Hello @PattaFeuFeu 
First, thanks for your work so far. Of course, your approach regarding KotlinX is totally fine. Honestly, I am not a Kotlin expert and thus was not aware of the existence of this functionality. But after checking out the docs a bit. I also think this is the right way to go. As you say, it seems better supported / integrated and also more lightweight I guess.  :-)

Regarding code style: Detekt for Kotlin code and PMD plus checkstyle for Java code should run once you run `gradle check`. 

Regarding your other questions, I will need to have a look into the code a bit more. Hopefully I can do this tomorrow.

**Comment by G10xy** [comment]
I am in complete agreement with @PattaFeuFeu. Kotlinx is much better and fits very good for the purposes of this library. By the way, terrific PR

**Comment by andregasser** [comment]
> Is my change in the test JSON data from Int to String for id-related properties okay? (see e.g. [Status API doc](https://docs.joinmastodon.org/entities/Status/) which mentions “Type: String (cast from an integer but not guaranteed to be a number)”
This is totally fine, even better! Quickly checked Mastodon API docs and that's actually how the data is returned from the server. So all good!! :-)

**Comment by andregasser** [comment]
> Is this a good way forward?

Definitively. Go ahead! Looking forward to your changes!

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt [hunk 1]
```diff
@@ -1,40 +1,13 @@
 package social.bigbone
 
-import com.google.gson.Gson
-import okhttp3.HttpUrl
-import okhttp3.Interceptor
+import kotlinx.serialization.json.Json
+import okhttp3.*
```
Comment (andregasser) [inline bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:4]: We discourage star imports in this project, so this should be changed. Actually, this should popup as a warning as well, if you run `gradle check`. Let me know if this is not the case.

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt [hunk 2]
```diff
@@ -736,3 +710,9 @@ private constructor(
         }
     }
 }
+
+private fun getConfiguredJson(): Json = Json {
```
Comment (andregasser) [inline bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:714]: This code appears in multiple places and could be deduplicated a bit I guess.

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt (hunk 1 above)
Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:4]: Ah, then that’s the first very obvious clash of the default style applied in IntelliJ. 😊 I’ll check.

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt (hunk 2 above)
Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:714]: See the “Still open” section in the PR description 😊

**Comment by andregasser** [comment]
I have gone through your code changes and must say, that this is looking very good so far. Looking forward to some test runs, once it is ready 🙂

File: bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt (hunk 1 above)
Comment (andregasser) [reply bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:4]: Eventually, I will try to create a config file for IntelliJ then. Thanks for pointing this out.

Comment (andregasser) [reply bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt:4]: I have created an issue for this. Let's collect all the deviations there: https://github.com/andregasser/bigbone/issues/279

**Comment by PattaFeuFeu** [comment]
@andregasser I think I’m mostly done. I will update the description soon and will add the pull request template to the bottom of the PR description with additional info.

How do we want to handle the decrease in code coverage? As all tests are still passing and I didn’t really add new functionality and rather replaced a dependency, I would personally ignore it (which also is the reason why I’m a bit sceptical of code coverage tools as “false positives” like this can happen easily).

If you’d like, I could potentially add some additional tests for edge cases. I introduced nullability in a few places where previously potential nullability was just ignored or covered up in a generic exception.

**Comment by andregasser** [comment]
> How do we want to handle the decrease in code coverage? As all tests are still passing and I didn’t really add new functionality and rather replaced a dependency, I would personally ignore it (which also is the reason why I’m a bit sceptical of code coverage tools as “false positives” like this can happen easily).

> If you’d like, I could potentially add some additional tests for edge cases. I introduced nullability in a few places where previously potential nullability was just ignored or covered up in a generic exception.

Do not worry about the test coverage tool output too much, as I am also not 100% happy with it yet. Because of this, I also did not set `codecov` as a mandatory check to be passed before we can merge changes. It is more of informational nature for the moment. Once I find more time, I'll have a look at it and try to improve / replace it.

I will do some test runs based on the changes in this PR very soon. 

Thanks for the update 🙂

File: bigbone/src/main/kotlin/social/bigbone/MastodonRequest.kt [hunk 3]
```diff
@@ -1,10 +1,9 @@
 package social.bigbone
 
-import com.google.gson.JsonParser
+import kotlinx.serialization.json.*
```
Comment (andregasser) [inline bigbone/src/main/kotlin/social/bigbone/MastodonRequest.kt:3]: Star import should be removed.

File: sample-kotlin/src/main/kotlin/social/bigbone/sample/GetInstanceInfo.kt [hunk 4]
```diff
@@ -1,9 +1,17 @@
 package social.bigbone.sample
 
-import com.google.gson.Gson
+import kotlinx.serialization.json.Json
 import social.bigbone.MastodonClient
+import social.bigbone.api.entity.Instance
 
 object GetInstanceInfo {
+
+    private val JSON_SERIALIZER: Json = Json {
```
Comment (andregasser) [inline sample-kotlin/src/main/kotlin/social/bigbone/sample/GetInstanceInfo.kt:9]: Can we use the same serializer config from `JsonSerializer.kt` eventually?

**Review by andregasser — CHANGES_REQUESTED** [review]
Overall, it looks very good, I have just added two minor comments. Thx!

File: sample-kotlin/src/main/kotlin/social/bigbone/sample/GetInstanceInfo.kt (hunk 4 above)
Comment (PattaFeuFeu) [reply sample-kotlin/src/main/kotlin/social/bigbone/sample/GetInstanceInfo.kt:9]: Depends on if we want to share it with consumers of this library. I was more of the opinion we should not and keep it private/an implementation detail.

What do you think? I know that GSON had been an `api`, and thus shared with consumers, dependency.

Comment (andregasser) [reply sample-kotlin/src/main/kotlin/social/bigbone/sample/GetInstanceInfo.kt:9]: Actually, the serializer should not be exposed to the consumer in any way, and thus be kept internal. I just wasn't sure if there's a way of sharing the serializer with the test classes without exposing it to the library consumer at ghe same time. 

Anway, your PR already is a big improvement, as we no longer expose the serializer to the consumer 👍🏻

We can keep this as it is. Fine for me 😊

Comment (PattaFeuFeu) [reply sample-kotlin/src/main/kotlin/social/bigbone/sample/GetInstanceInfo.kt:9]: Not as far as I’m aware as, in the end, the sample apps are mere consumers of the library as well and would thus receive the same shared dependencies as any other consumer would. 🤔

File: bigbone/build.gradle [hunk 5]
```diff
@@ -1,16 +1,19 @@
 plugins {
     id 'bigbone.library-conventions'
     id 'bigbone.integrationtest-conventions'
+    alias(libs.plugins.kotlin.serialization)
 }
 
 dependencies {
     api libs.okhttp
-    api libs.gson
+    api libs.kotlinx.serialization.json
```
Comment (PattaFeuFeu) [inline bigbone/build.gradle:9]: Ah. Dang. I still had this as api here. 🙈

Comment (PattaFeuFeu) [reply bigbone/build.gradle:9]: Personally, I’d change this to `implementation` per my reasoning in the other comments. 

@andregasser Feel free to update this PR if you can. I won’t be near a computer until tomorrow likely but would like this to get merged soon.

Comment (PattaFeuFeu) [reply bigbone/build.gradle]: @andregasser I have changed this to `implementation` in ae8ab1c and have also removed the dependency on kotlinx serialization from both sample modules as they only used it for a log output. I’ve replaced that one with a simple `toString` which, as it is a `data class`, still provides helpful output just for testing the sample.

**Review by bocops — DISMISSED** [review]
This looks good to me! Regarding some of the mostly code style questions brought up earlier:

- changing id-related properties in test data to string: this is correct. We've had to change all id types to strings earlier this year, the test data is probably mostly left over from before that change.
- using argument names: I don't think it has ever been a conscious choice not to use them. I prefer them for clarity as well, so as far as I am concerned, we should keep it this way and use argument names throughout.

**Comment by andregasser** [comment]
> This looks good to me! Regarding some of the mostly code style questions brought up earlier:
> 
> * changing id-related properties in test data to string: this is correct. We've had to change all id types to strings earlier this year, the test data is probably mostly left over from before that change.
> * using argument names: I don't think it has ever been a conscious choice not to use them. I prefer them for clarity as well, so as far as I am concerned, we should keep it this way and use argument names throughout.

Fine for me as well. Let's use argument names in a consistent way.

**Review by andregasser — DISMISSED** [review]

**Review by andregasser — APPROVED** [review]
Ok!
